File: Rock.Obsidian/Blocks/Crm/AttributeValues.cs
// <copyright>
// Copyright by the Spark Development Network
//
// Licensed under the Rock Community License (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.rockrms.com/license
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// </copyright>
//
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Net;
using Rock.Attribute;
using Rock.Blocks;
using Rock.Data;
using Rock.Model;
using Rock.Web.Cache;

namespace Rock.Obsidian.Blocks.Crm
{
    /// <summary>
    /// Allows for editing the value(s) of a set of attributes for person.
    /// </summary>
    /// <seealso cref="Rock.Blocks.ObsidianBlockType" />

    [DisplayName( "Attribute Values" )]
    [Category( "Obsidian > CRM" )]
    [Description( "Allows for editing the value(s) of a set of attributes for person." )]
    [IconCssClass( "fa fa-user" )]

    #region Block Attributes

    [AttributeCategoryField(
        "Category",
        Key = AttributeKey.Category,
        AllowMultiple = true,
        Description = "The Attribute Categories to display attributes from",
        EntityTypeName = "Rock.Model.Person",
        IsRequired = true,
        Order = 0 )]

    [TextField(
        "Attribute Order",
        Key = AttributeKey.AttributeOrder,
        Description = "The order to use for displaying attributes.  Note: this value is set through the block's UI and does not need to be set here.",
        IsRequired = false,
        Order = 1 )]

    [BooleanField(
        "Use Abbreviated Name",
        Key = AttributeKey.UseAbbreviatedName,
        Description = "Display the abbreviated name for the attribute if it exists, otherwise the full name is shown.",
        IsRequired = true,
        DefaultBooleanValue = false,
        Order = 2 )]

    [TextField( "Block Title",
        Key = AttributeKey.BlockTitle,
        Description = "The text to display as the heading.",
        IsRequired = false,
        DefaultValue = "",
        Order = 3 )]

    [TextField(
        "Block Icon",
        Key = AttributeKey.BlockIcon,
        Description = "The css class name to use for the heading icon.",
        IsRequired = false,
        DefaultValue = "",
        Order = 4 )]

    [BooleanField(
        "Show Category Names as Separators",
        Key = AttributeKey.ShowCategoryNamesAsSeparators,
        Description = "If enabled, attributes will be grouped by category and will include the category name as a heading separator.",
        IsRequired = true,
        DefaultBooleanValue = false,
        Order = 5 )]

    #endregion Block Attributes

    public class AttributeValues : ObsidianBlockType
    {
        #region Attribute Keys
        private static class AttributeKey
        {
            public const string Category = "Category";
            public const string AttributeOrder = "AttributeOrder";
            public const string UseAbbreviatedName = "UseAbbreviatedName";
            public const string BlockTitle = "BlockTitle";
            public const string BlockIcon = "BlockIcon";
            public const string ShowCategoryNamesAsSeparators = "ShowCategoryNamesasSeparators";
        }

        #endregion Attribute Keys

        #region IObsidianBlockType Implementation

        /// <summary>
        /// Gets the property values that will be sent to the browser.
        /// </summary>
        /// <returns>
        /// A collection of string/object pairs.
        /// </returns>
        public override object GetConfigurationValues()
        {
            return new
            {
                BlockIconCssClass = GetBlockIconCssClass(),
                BlockTitle = GetBlockTitle(),
                ShowCategoryNamesAsSeparators = GetAttributeValue( AttributeKey.ShowCategoryNamesAsSeparators ).AsBoolean()
            };
        }

        #endregion

        #region Actions

        /// <summary>
        /// Get data based on the configured category setting.
        /// </summary>
        [BlockAction( "GetAttributeDataList" )]
        public BlockActionResult GetAttributeDataList()
        {
            var currentPerson = GetCurrentPerson();
            var categories = GetCategoryGuids();
            var attributeDataList = new List<AttributeData>();

            if ( !categories.Any() || currentPerson == null )
            {
                return new BlockActionResult( HttpStatusCode.OK, attributeDataList );
            }

            var rockContext = new RockContext();
            var attributeService = new AttributeService( rockContext );
            var orderOverride = GetAttributeValue( AttributeKey.AttributeOrder ).SplitDelimitedValues().AsIntegerList();

            var orderedAttributeList = attributeService.Queryable()
                .Where( a => a.IsActive && a.Categories.Any( c => categories.Contains( c.Guid ) ) )
                .OrderBy( a => a.Order )
                .ThenBy( a => a.Name )
                .ToAttributeCacheList();

            foreach ( var attributeId in orderOverride )
            {
                var attribute = orderedAttributeList.FirstOrDefault( a => a.Id == attributeId );

                if ( attribute != null && attribute.IsAuthorized( Rock.Security.Authorization.VIEW, currentPerson ) )
                {
                    attributeDataList.Add( GetAttributeData( attribute ) );
                }
            }

            foreach ( var attribute in orderedAttributeList.Where( a => !orderOverride.Contains( a.Id ) ) )
            {
                if ( attribute.IsAuthorized( Rock.Security.Authorization.VIEW, currentPerson ) )
                {
                    attributeDataList.Add( GetAttributeData( attribute ) );
                }
            }

            attributeDataList = attributeDataList.Where( a => a != null ).ToList();
            return new BlockActionResult( HttpStatusCode.OK, attributeDataList );
        }

        /// <summary>
        /// Gets the attribute data.
        /// </summary>
        /// <param name="attribute">The attribute.</param>
        /// <returns></returns>
        private AttributeData GetAttributeData( AttributeCache attribute )
        {
            var useAbbreviatedName = GetAttributeValue( AttributeKey.UseAbbreviatedName ).AsBoolean();
            var contextPerson = GetContextPerson();
            var attributeValue = contextPerson.GetAttributeValue( attribute.Key );

            if ( attributeValue.IsNullOrWhiteSpace() )
            {
                return null;
            }

            var data = new AttributeData {
                Label = useAbbreviatedName ? attribute.AbbreviatedName : attribute.Name,
                Value = attributeValue,
                FieldTypeGuid = attribute.FieldType.Guid
            };

            return data;
        }

        #endregion Actions

        #region Data Access

        /// <summary>
        /// Gets the block icon CSS class.
        /// </summary>
        /// <returns></returns>
        private string GetBlockIconCssClass()
        {
            var icon = GetAttributeValue( AttributeKey.BlockIcon );

            if (!icon.IsNullOrWhiteSpace())
            {
                return icon;
            }

            var category = GetSoloCategory();

            if ( category != null )
            {
                return category.IconCssClass;
            }

            return "fa fa-user";
        }

        /// <summary>
        /// Gets the block title.
        /// </summary>
        /// <returns></returns>
        private string GetBlockTitle()
        {
            var blockTitle = GetAttributeValue( AttributeKey.BlockTitle );

            if ( !blockTitle.IsNullOrWhiteSpace() )
            {
                return blockTitle;
            }

            var category = GetSoloCategory();

            if ( category != null )
            {
                return category.Name;
            }

            return "Attribute Values";
        }

        /// <summary>
        /// Gets the solo category.
        /// </summary>
        /// <returns></returns>
        private CategoryCache GetSoloCategory()
        {
            var categoryGuids = GetCategoryGuids();
            return categoryGuids.Count == 1 ? CategoryCache.Get( categoryGuids[0] ) : null;
        }

        /// <summary>
        /// Gets the category Guids.
        /// </summary>
        /// <returns></returns>
        private List<Guid> GetCategoryGuids()
        {
            return GetAttributeValue( AttributeKey.Category ).SplitDelimitedValues( false ).AsGuidList();
        }

        /// <summary>
        /// Gets the context person.
        /// </summary>
        /// <returns></returns>
        private Person GetContextPerson()
        {
            if ( _contextPerson != null )
            {
                return _contextPerson;
            }

            var rockContext = new RockContext();
            var personService = new PersonService( rockContext );

            _contextPerson = personService.Get( 4 );
            _contextPerson?.LoadAttributes();

            return _contextPerson;
        }
        private Person _contextPerson = null;

        #endregion Data Access

        #region View Models

        /// <summary>
        /// A login result object
        /// </summary>
        public class AttributeData
        {
            /// <summary>
            /// Gets or sets the label.
            /// </summary>
            public string Label { get; set; }

            /// <summary>
            /// Gets or sets the formatted value.
            /// </summary>
            public string Value { get; set; }

            /// <summary>
            /// Gets the field type unique identifier.
            /// </summary>
            public Guid FieldTypeGuid { get; set; }
        }

        #endregion View Models
    }
}

File: Rock.JavaScript.Obsidian/Framework/Blocks/Crm/attributeValues.ts
// <copyright>
// Copyright by the Spark Development Network
//
// Licensed under the Rock Community License (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.rockrms.com/license
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// </copyright>
//
import { computed, defineComponent, ref } from "vue";
import PaneledBlockTemplate from "../../Templates/paneledBlockTemplate";
import Loading from "../../Controls/loading";
import Alert from "../../Elements/alert";
import { useStore } from "../../Store/index";
import { Guid } from "../../Util/guid";
import { useConfigurationValues, useInvokeBlockAction } from "../../Util/block";
import JavaScriptAnchor from "../../Elements/javaScriptAnchor";
import RockForm from "../../Controls/rockForm";
import TextBox from "../../Elements/textBox";
import RockButton from "../../Elements/rockButton";
import { PublicAttributeValue, PublicEditableAttributeValue } from "../../ViewModels";
import AttributeValuesContainer from "../../Controls/attributeValuesContainer";
import { List } from "../../Util/linq";

const store = useStore();

type ConfigurationValues = {
    blockIconCssClass: string;

    blockTitle: string;

    showCategoryNamesAsSeparators: boolean;

    useAbbreviatedNames: boolean;

    categoryGuids: Guid[];

    attributes: PublicAttributeValue[];
};

function sortedAttributeValues(attributeValues: PublicAttributeValue[]): PublicAttributeValue[] {
    return new List(attributeValues)
        .orderBy(v => v.order)
        .thenBy(v => v.name)
        .toArray();
}

export default defineComponent({
    name: "Crm.AttributeValues",
    components: {
        PaneledBlockTemplate,
        Alert,
        Loading,
        JavaScriptAnchor,
        RockForm,
        TextBox,
        RockButton,
        AttributeValuesContainer
    },
    setup() {
        const configurationValues = useConfigurationValues<ConfigurationValues>();
        const invokeBlockAction = useInvokeBlockAction();
        const attributeValues = ref(sortedAttributeValues(configurationValues.attributes));
        const personGuid = computed(() => store.personContext?.guid || null);
        const isLoading = ref(false);
        const isEditMode = ref(false);
        const errorMessage = ref("");

        const goToViewMode = (): void => {
            isEditMode.value = false;
        };

        const goToEditMode = async (): Promise<void> => {
            const result = await invokeBlockAction<PublicEditableAttributeValue[]>("GetAttributeValuesForEdit");
            if (result.isSuccess) {
                attributeValues.value = sortedAttributeValues(result.data ?? []);
                isEditMode.value = true;
            }
        };

        const doSave = async (): Promise<void> => {
            isLoading.value = true;

            const keyValueMap: Record<string, string | null> = {};

            for (const a of attributeValues.value) {
                keyValueMap[(a as PublicEditableAttributeValue).key] = a.value || "";
            }

            const result = await invokeBlockAction<PublicAttributeValue[]>("SaveAttributeValues", {
                personGuid: personGuid.value,
                keyValueMap
            });

            if (result.isSuccess) {
                attributeValues.value = sortedAttributeValues(result.data ?? []);
                goToViewMode();
            }
            else {
                errorMessage.value = "Failed to save values.";
            }

            isLoading.value = false;
        };

        return {
            blockTitle: computed(() => configurationValues.blockTitle),
            blockIconCssClass: computed(() => configurationValues.blockIconCssClass),
            isLoading,
            isEditMode,
            errorMessage,
            goToViewMode,
            goToEditMode,
            doSave,
            useAbbreviatedNames: configurationValues.useAbbreviatedNames,
            attributeValues
        };
    },
    template: `
<PaneledBlockTemplate class="panel-persondetails">
    <template v-slot:title>
        <i :class="blockIconCssClass"></i>
        {{ blockTitle }}
    </template>
    <template v-slot:titleAside>
        <div class="actions rollover-item pull-right">
            <JavaScriptAnchor title="Order Attributes" class="btn-link edit">
                <i class="fa fa-bars"></i>
            </JavaScriptAnchor>
            <JavaScriptAnchor title="Edit Attributes" class="btn-link edit" @click="goToEditMode">
                <i class="fa fa-pencil"></i>
            </JavaScriptAnchor>
        </div>
    </template>
    <template v-slot:default>
        <Loading :isLoading="isLoading">
            <Alert v-if="errorMessage" alertType="warning">{{ errorMessage }}</Alert>
            <AttributeValuesContainer v-if="!isEditMode" :attributeValues="attributeValues" :showEmptyValues="false" />
            <RockForm v-else @submit="doSave">
                <AttributeValuesContainer :attributeValues="attributeValues" isEditMode :showAbbreviatedName="useAbbreviatedNames" />
                <div class="actions">
                    <RockButton btnType="primary" btnSize="xs" type="submit">Save</RockButton>
                    <RockButton btnType="link" btnSize="xs" @click="goToViewMode">Cancel</RockButton>
                </div>
            </RockForm>
        </Loading>
    </template>
</PaneledBlockTemplate>`
});
